starknet/src/utils/eip712.cairo

```diff
@@ -0,0 +1,201 @@
+    metadata_URI: Span<felt252>,
+    salt: u256
+) {
+    let digest: u256 = get_update_proposal_digest(
+        domain_hash, target, author, proposal_id, execution_strategy, metadata_URI, salt
+    );
+    verify_eth_signature::<Secp256k1Point>(digest, signature_from_vrs(v, r, s), author);
+}
+
+fn get_propose_digest(
+    domain_hash: u256,
+    space: ContractAddress,
+    author: EthAddress,
+    execution_strategy: @Strategy,
+    user_proposal_validation_params: Span<felt252>,
+    metadata_URI: Span<felt252>,
+    salt: u256
+) -> u256 {
+    let encoded_data = array![
+        u256 { low: PROPOSE_TYPEHASH_LOW, high: PROPOSE_TYPEHASH_HIGH },
+        get_contract_address().into(),
+        space.into(),
+        author.into(),
+        execution_strategy.keccak_struct_hash(),
+        user_proposal_validation_params.keccak_struct_hash(),
+        metadata_URI.keccak_struct_hash(),
+        salt
+    ];
+    let message_hash = keccak::keccak_u256s_be_inputs(encoded_data.span()).byte_reverse();
+    hash_typed_data(domain_hash, message_hash)
+}
+
+fn get_vote_digest(
+    domain_hash: u256,
+    space: ContractAddress,
+    voter: EthAddress,
+    proposal_id: u256,
+    choice: Choice,
+    user_voting_strategies: Span<IndexedStrategy>,
+    metadata_URI: Span<felt252>,
+) -> u256 {
+    let encoded_data = array![
+        u256 { low: VOTE_TYPEHASH_LOW, high: VOTE_TYPEHASH_HIGH },
+        get_contract_address().into(),
+        space.into(),
+        voter.into(),
+        proposal_id,
+        choice.into(),
+        user_voting_strategies.keccak_struct_hash(),
+        metadata_URI.keccak_struct_hash()
+    ];
+    let message_hash = keccak::keccak_u256s_be_inputs(encoded_data.span()).byte_reverse();
+    hash_typed_data(domain_hash, message_hash)
+}
+
+fn get_update_proposal_digest(
+    domain_hash: u256,
+    space: ContractAddress,
+    author: EthAddress,
+    proposal_id: u256,
+    execution_strategy: @Strategy,
+    metadata_URI: Span<felt252>,
+    salt: u256
+) -> u256 {
+    let encoded_data = array![
+        u256 { low: UPDATE_PROPOSAL_TYPEHASH_LOW, high: UPDATE_PROPOSAL_TYPEHASH_HIGH },
+        get_contract_address().into(),
+        space.into(),
+        author.into(),
+        proposal_id,
+        execution_strategy.keccak_struct_hash(),
+        metadata_URI.keccak_struct_hash(),
+        salt
+    ];
+    let message_hash = keccak::keccak_u256s_be_inputs(encoded_data.span()).byte_reverse();
+    hash_typed_data(domain_hash, message_hash)
+}
+
+fn get_domain_hash() -> u256 {
+    // The ethers typed data encoder is not compatible with a Starknet address as the `verifyingContract`
+    // therefore we cannot use the `verifyingContract` field in the domain separator, instead we add the 
+    //  verifying contract address to the message itself.
+    let encoded_data = array![
+        u256 { low: DOMAIN_TYPEHASH_LOW, high: DOMAIN_TYPEHASH_HIGH },
+        Felt252IntoU256::into(get_tx_info().unbox().chain_id)
```

Comment:
`get_tx_info().unbox().chaind_id.into()` would work also I think?

Reply:
Since there is a core lib `Felt252IntoU256` but then we also have the generic `TIntoU256` impl there is 2 impls in scope for the same thing. So need to specify which one to use.

Reply:
oh right :) this is one of the quirks of `impl ... of ...` I guess :p (rust doesn't have this issue :p)